src/steamship/agents/service/agent_service.py

```diff
@@ -131,15 +131,48 @@ def run_action(self, agent: Agent, action: Action, context: AgentContext):
                 context.action_cache.update(key=action, value=action.output)
 
     def run_agent(self, agent: Agent, context: AgentContext):
-        # first, clear any prior agent steps from set of completed steps
-        # this will allow the agent to select tools/dispatch actions based on a new context
+        """Run the agent.
+
+        This is the root method that runs when the agent awakes to do work. Any chat messages to response to should
+        have already been appended to the ChatHistory associated with the AgentContext.
+        """
+
+        # Clear any prior agent steps from set of completed steps.
+        # This will allow the agent to select tools/dispatch actions based on a new context
         context.completed_steps = []
 
-        action = self.next_action(
-            agent=agent, input_blocks=[context.chat_history.last_user_message], context=context
-        )
+        last_user_message = context.chat_history.last_user_message
+
+        if last_user_message is None:
+            # There is nothing to respond to.
+            # TODO: Consider the situations in which an agent awakes for reasons other than message response.
+            logging.info(
+                "AgentService.run_agent invoked but no last user message existed.",
+                extra={
+                    AgentLogging.IS_MESSAGE: False,
+                    AgentLogging.MESSAGE_AUTHOR: AgentLogging.AGENT,
+                },
+            )
+            return
+
+        if last_user_message.response_sent:
```

Comment:
does this imply that, from an agent perspective, a single user message can ONLY have a singular response (perhaps consisting of multiple blocks)?  There is no response-in-progress type state?

Reply:
```
"does this imply that, from an agent perspective, a single user message can ONLY have a singular response (perhaps consisting of multiple blocks)?"
```

Yeah that is the semantics this implements, except for this edge case below. Does that block a usage scenario we've got?

```
There is no response-in-progress type state?
```

Yeah there's a `delta-t` in between `webhook-received` and `last-message-marked-as-responded-to` in which the agent could always double respond. As implemented here, that `doublt-t` spans the duration of the entire response generation. We could also shorten it by having an `response-in-progress` state that is more quickly added to the message.

Do you think that's important to add here?

Reply:
I don't know that it is a blocker for any particular use case we have. I'd have to invent a scenario in which having multiple agents respond to a single user input is the "right" thing.

In general, though, I'm wondering if this solves the core issue that was being observed, and how that manifests.

Is the issue that the agent is being triggered to run on the _same_ block/message?  Or that a new, *identical* block/message is being added because a client is *resending* the same content?

Because if it is the first scenario, I want to understand _why_ that is happening and what the implications are for latency and in-flight state.

If it is the second, I'm not sure `response_sent` solves the issue.